xds/internal/server/rds_handler_test.go

```diff
@@ -0,0 +1,423 @@
+	// closed, and the call should cancel all the watches present (for this
+	// test, two watches on route1 and route2).
+	rh.close()
+	routeNameDeleted, err := fakeClient.WaitForCancelRouteConfigWatch(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.CancelRDS failed with error: %v", err)
+	}
+	if routeNameDeleted != route1 && routeNameDeleted != route2 {
+		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v or %v", routeNameDeleted, route1, route2)
+	}
+
+	routeNameDeleted, err = fakeClient.WaitForCancelRouteConfigWatch(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.CancelRDS failed with error: %v", err)
+	}
+	if routeNameDeleted != route1 && routeNameDeleted != route2 {
+		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v or %v", routeNameDeleted, route1, route2)
+	}
+}
+
+// TestSuccessCaseDeletedRoute tests the case where the rds handler receives an
+// update with two routes, then receives an update with only one route. The RDS
+// Handler is expected to cancel the watch for the route no longer present.
+func (s) TestSuccessCaseDeletedRoute(t *testing.T) {
+	rh, fakeClient := setupTests(t)
+
+	routeNames := map[string]bool{route1: true, route2: true}
+	rh.updateRouteNamesToWatch(routeNames)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Will start two watches.
+	_, err := fakeClient.WaitForWatchRouteConfig(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchRDS failed with error: %v", err)
+	}
+	_, err = fakeClient.WaitForWatchRouteConfig(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchRDS failed with error %v", err)
+	}
+
+	// Update the RDSHandler with route names which deletes a route name to
+	// watch. This should trigger the RDSHandler to cancel the watch for the
+	// deleted route name to watch.
+	routeNames = map[string]bool{route1: true}
+	rh.updateRouteNamesToWatch(routeNames)
+	// This should delete the watch for route2.
+	routeNameDeleted, err := fakeClient.WaitForCancelRouteConfigWatch(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.CancelRDS failed with error %v", err)
+	}
+	if routeNameDeleted != route2 {
+		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v", routeNameDeleted, route2)
+	}
+
+	rdsUpdate := xdsclient.RouteConfigUpdate{}
+	// Invoke callback with the xds client with a certain route update. Due to
+	// this route update updating every route name that rds handler handles,
+	// this should write to the update channel to send to the listener.
+	fakeClient.InvokeWatchRouteConfigCallback(route1, rdsUpdate, nil)
+	rhuWant := map[string]xdsclient.RouteConfigUpdate{route1: rdsUpdate}
+	select {
+	case rhu := <-rh.updateChannel:
+		if diff := cmp.Diff(rhu.updates, rhuWant); diff != "" {
+			t.Fatalf("got unexpected route update, diff (-got, +want): %v", diff)
+		}
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for update from update channel.")
+	}
+
+	rh.close()
+	_, err = fakeClient.WaitForCancelRouteConfigWatch(ctx)
```

Comment:
Same here, need to ensure that the correct name is watched for.

One idea for a helper with variadic args is as follows:
```golang
func waitForFuncWithNames(ctx context.Context, f func(ctx) (string, err), names ...string) error {
  wantNames := make(map[string]bool, len(names))
  for _, name := range names {
    wantNames[name] = true
  }
  gotNames := make(map[string]bool, len(names))
  for name := range wantNames {
    name, err := f(ctx)
    if err != nil {
      return err
    }
    gotNames[name] = true
  }
  if !cmp.Equal(gotNames, wantNames) {
    return fmt.Errorf("got routeNames %v, want %v", gotNames, wantNames)
  }
}

Reply:
Implemented your helper. Plumbed it everywhere with nondeterminism. (Made sure it was working by making sure it failed if I changed variadic names to something incorrect).